### PR TITLE
Cooperate with custom error handlers

### LINKED ISSUE #2276: V3 Handler not shutting up after custom error handler.
# Other bugs

#### What were you trying to do?

Replace a redbot error message (in this case the message displayed after CommandOnCooldown was triggered) with a custom message via @my_function.error:

    @my_function.error
    async def myfunction_error(self, ctx, error):
        if isinstance(error, commands.CommandOnCooldown):
             await ctx.send("I feel a little tired now. Ask me again in {:.0f}s".format(error.retry_after))

#### What were you expecting to happen?

To only see my custom on_cooldown message from the event above.

#### What actually happened?

The bot wrote:
I feel a little tired now. Ask me again in 117s
This command is on cooldown. Try again in 116.69s

#### How can we reproduce this issue?

Add cooldown to any command and extend the @functionname.error handler like I described for your command. Then cause the command cooldown and try to use the command again while on cooldown.

### PR DESCRIPTION
**Pull request by DiscordLiz**
Adds detection of custom error handlers to the global error handler.

Adds documentation of the updated behavior.

closes #2276

### Type

- [x] Bugfix
- [ ] Enhancement
- [ ] New feature

### CODE REVIEW

**Review by calebj — CHANGES_REQUESTED** [review]
It looks like Red already ignores (or at least, doesn't post) errors if a cog-wide exception handler is present (see line 191). If that becomes unreachable code, better to just skip that check altogether.

**Comment by DiscordLiz** [comment]
I handled the feedback brought up on Discord and on GitHub separately.

@calebj brought up in discord that cog authors may not want to handle every error red does themselves.

I added a keyword only argument in the latest commit to allow developers to pass their errors back to Red's handler and have it skip the checks for if there were a cog handler already.

Other solutions to this which were discussed in discord.

1. Split error handling up into another function and the event. (@mikeshardmind brought this up)

This would incure an additional function call per error which wasn't handled, but would leave the event signature untouched.

2. Move error handling into the cog base class. (@mikeshardmind brought this up)

This would require significant changes to error handling to select the first handler in a cog class's mro, but would allow passing of unhandled exceptions using super.

3. Create a new convention of a red specific error handler for each type of error. (@calebj brought this up)

This would put an additional learning curve to new developers transitioning from using discord.py without Red, and would still require significant changes to upstream code to prevent double error handling.

This would also not allow people to use a cog error handler to log exceptions broadly without duplicating code.

**Review by calebj — CHANGES_REQUESTED** [review]
Is it `manually_recalled` or `manually_reinvoked`?

**Review by calebj — DISMISSED** [review]

File: redbot/core/commands/commands.py [hunk 1]
```diff
@@ -398,6 +398,44 @@ def clear_rule_for(
...
+Parameters
+-----------
+coro : :ref:`coroutine <coroutine>`
+    The coroutine to register as the local error handler.
+
+Raises
+-------
+discord.ClientException
+    The coroutine is not actually a coroutine.
+"""
```
Comment (Tobotimus) [inline redbot/core/commands/commands.py]: Could you please create the `error(self, coro)` method on our command class to just call `super().error(coro)`, and then move this docstring there? That way it'll show up in our docs, and is more obvious in the code.

**Review by Tobotimus — CHANGES_REQUESTED** [review]
Just one change I'm requesting then it's good :)

File: redbot/core/commands/commands.py (hunk 1 above)
Comment (DiscordLiz) [reply redbot/core/commands/commands.py]: I've created the method and moved the docstring.

**Review by Tobotimus — APPROVED** [review]

**Pull request merged**